quinn/src/tests.rs

```diff
@@ -183,70 +195,88 @@ fn export_keying_material() {
-        .await
-        .expect("connection");
-
-    // ...and read what was sent.
-    let mut stream = receiver.accept_uni().await.expect("incoming streams");
-    let msg = stream
-        .read_to_end(usize::max_value())
-        .await
-        .expect("read_to_end");
-    assert_eq!(msg, MSG);
-
-    // But it's still definitely closed.
-    assert!(receiver.open_uni().await.is_err());
+    let endpoint_factory = EndpointFactory::new();
+    let client_1 = endpoint_factory.endpoint();
+    let client_1_addr = client_1.local_addr().unwrap();
+    let client_2 = endpoint_factory.endpoint();
+    let server = endpoint_factory.endpoint();
+    let server_addr = server.local_addr().unwrap();
+    let server_task = tokio::spawn(async move {
+        loop {
+            let accepting = server.accept().await.unwrap();
+            if accepting.remote_address() == client_1_addr {
+                accepting.reject();
+            } else if accepting.is_validated() {
+                accepting.await.expect("connection");
+            } else {
+                accepting.retry();
+            }
+        }
+    });
+    let client_1_task = tokio::spawn(async move {
+        client_1
+            .connect(server_addr, "localhost")
+            .unwrap()
+            .await
+            .expect_err("server should have blocked this");
+    });
+    let client_2_task = tokio::spawn(async move {
+        client_2
+            .connect(server_addr, "localhost")
+            .unwrap()
+            .await
+            .expect("connect");
+    });
+    client_1_task.await.unwrap();
+    client_2_task.await.unwrap();
+    server_task.abort();
 }
 
 /// Construct an endpoint suitable for connecting to itself
 fn endpoint() -> Endpoint {
-    endpoint_with_config(TransportConfig::default())
+    EndpointFactory::new().endpoint()
 }
 
 fn endpoint_with_config(transport_config: TransportConfig) -> Endpoint {
-    let cert = rcgen::generate_simple_self_signed(vec!["localhost".into()]).unwrap();
-    let key = rustls::PrivateKey(cert.serialize_private_key_der());
-    let cert = rustls::Certificate(cert.serialize_der().unwrap());
-    let transport_config = Arc::new(transport_config);
-    let mut server_config = crate::ServerConfig::with_single_cert(vec![cert.clone()], key).unwrap();
-    server_config.transport_config(transport_config.clone());
+    EndpointFactory::new().endpoint_with_config(transport_config)
+}
 
-    let mut roots = rustls::RootCertStore::empty();
-    roots.add(&cert).unwrap();
-    let mut endpoint = Endpoint::server(
-        server_config,
-        SocketAddr::new(IpAddr::V4(Ipv4Addr::LOCALHOST), 0),
-    )
-    .unwrap();
-    let mut client_config = ClientConfig::with_root_certificates(roots);
-    client_config.transport_config(transport_config);
-    endpoint.set_default_client_config(client_config);
+/// Constructs endpoints suitable for connecting to themselves and each other
+struct EndpointFactory(rcgen::Certificate);
```

Comment:
What's the purpose of this refactoring? It's not obviously required by any other changes in this commit.

Reply:
The new `ip_blocking` test creates 3 endpoints, specifically relying on the 2 that are designated client endpoint having different IP addresses, and tests the behavior of the server blocking one of their IP addresses and not the other. `EndpointFactory` is factored out so that their different endpoints can be constructed with compatible keys, which tests weren't previously doing. Creating multiple endpoints is necessary to give them different IP addresses, which other tests haven't been doing.